### PR TITLE
[Module Aliasing] Update alias tracker / handle edge cases

### PR DESCRIPTION
**Pull request by elsh**
Update module alias tracker 
Move module aliases setting per target into alias tracker 
Improve perf
Handle edge cases
Add more tests
Resolves rdar://90337780

### CODE REVIEW

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@elsh is this a cherry pick from the main branch?

**Comment by elsh** [comment]
> ~@elsh is this a cherry pick from the main branch?~
> 
> @elsh this is for the main branch but marked as 5.7, is this intentional?

oops, it's for main; updated the title/label.

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@elsh since this is a fairly narrowly specialized change, any specific areas in the change-set that you would like to highlight for review?

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 1]
```diff
@@ -728,29 +695,22 @@ private class ModuleAliasTracker {
...
+    func trackTargetsPerProduct(_ targets: [Target],
+                                product: String,
+                                package: PackageIdentity) {
+        // This should be called once per product
+        idProductToTargets[package, default: [:]][product] = targets
     }
 
-    func validateSources(_ targets: [Target],
-                         product: String,
+    func validateAndApplyAliases(product: String,
```
Comment (tomerd) [inline Sources/PackageGraph/PackageGraph+Loading.swift:705]: nit: formatting

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 2]
```diff
@@ -728,29 +695,22 @@ private class ModuleAliasTracker {
...
                          package: PackageIdentity) throws {
+        guard let targets = idProductToTargets[package]?[product] else { return }
         for target in targets {
-            if let aliases = idTargetToAliases[package]?[target.name], !aliases.isEmpty {
+            if target.moduleAliases != nil {
```
Comment (tomerd) [inline Sources/PackageGraph/PackageGraph+Loading.swift:709]: nit, filter at the loop level?

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 3]
```diff
@@ -764,34 +724,54 @@ private class ModuleAliasTracker {
...
-                                }
-                            }
+        var buffer = [String: ModuleAliasModel]()
+        propagate(at: rootPkg, using: &buffer)
+    }
+
+    func propagate(at cur: PackageIdentity, using buffer: inout [String: ModuleAliasModel]) {
+        if let productToTargetAliases = aliasMap[cur] {
+            for (_, aliasModels) in productToTargetAliases {
+                for aliasModel in aliasModels {
```
Comment (tomerd) [inline Sources/PackageGraph/PackageGraph+Loading.swift:735]: nit: flatten loops

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 4]
```diff
@@ -764,34 +724,54 @@ private class ModuleAliasTracker {
...
-                                    for childModel in childAliases {
-                                        childModel.alias = parentModel.alias
-                                    }
-                                }
-                            }
+        var buffer = [String: ModuleAliasModel]()
+        propagate(at: rootPkg, using: &buffer)
+    }
+
+    func propagate(at cur: PackageIdentity, using buffer: inout [String: ModuleAliasModel]) {
```
Comment (tomerd) [inline Sources/PackageGraph/PackageGraph+Loading.swift:732]: use full qualified variable names in method signatures

**Comment by elsh** [comment]
@swift-ci smoke test

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 5]
```diff
@@ -199,32 +199,6 @@ private func checkAllDependenciesAreUsed(_ rootPackages: [ResolvedPackage], obse
     }
 }
 
-extension Package {
-    // Add module aliases specified for applicable targets
-    fileprivate func setModuleAliasesForTargets(with moduleAliasMap: [String: [ModuleAliasModel]]) {
```
Comment (elsh) [inline Sources/PackageGraph/PackageGraph+Loading.swift:204]: this is removed / setting is now done in the alias tracker

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 6]
```diff
@@ -728,30 +695,22 @@ private class ModuleAliasTracker {
...
-                }
+    func validateAndApplyAliases(product: String,
+                                 package: PackageIdentity) throws {
+        guard let targets = idProductToTargets[package]?[product] else { return }
+        let targetsWithAliases = targets.filter{ $0.moduleAliases != nil }
+        for target in targetsWithAliases {
+            if target.sources.containsNonSwiftFiles {
+                throw PackageGraphError.invalidSourcesForModuleAliasing(target: target.name, product: product, package: package.description)
             }
+            target.applyAlias()
```
Comment (elsh) [inline Sources/PackageGraph/PackageGraph+Loading.swift:713]: aliasing/renaming target is done here now

File: Tests/BuildTests/BuildPlanTests.swift [hunk 7]
```diff
@@ -2359,6 +2359,167 @@ final class BuildPlanTests: XCTestCase {
...
+                    dependencies: [
+                        .localSourceControl(path: .init("/aPkg"), requirement: .upToNextMajor(from: "1.0.0")),
+                        .localSourceControl(path: .init("/xPkg"), requirement: .upToNextMajor(from: "1.0.0")),
+                    ],
+                    targets: [
+                        TargetDescription(name: "App",
+                                          dependencies: [ "Foo",
+                                                         .product(name: "A",
+                                                                  package: "aPkg",
+                                                                  moduleAliases: ["Foo": "FooA"]),
```
Comment (elsh) [inline Tests/BuildTests/BuildPlanTests.swift:2489]: the updated propagation method allows aliasing targets that are multi-level upstream at the user level

**Comment by elsh** [comment]
> @elsh since this is a fairly narrowly specialized change, any specific areas in the change-set that you would like to highlight for review?

highlighted inline in comments

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 8]
```diff
@@ -634,27 +603,22 @@ private func resolveModuleAliases(packageBuilders: [ResolvedPackageBuilder],
...
-                                                      consumingPackage: packageBuilder.package.identity)
-                        }
-                    }
+        let targetDependencies = packageBuilder.package.targets.map {$0.dependencies}.flatMap{$0}
+        for dep in targetDependencies {
+            if case let .product(productRef, _) = dep,
+               let productPkg = productRef.package {
+                let productPkgID = PackageIdentity.plain(productPkg)
+                // Track dependency package ID chain
+                aliasTracker.addPackageIDChain(parent: packageBuilder.package.identity, child: productPkgID)
```
Comment (tomerd) [inline Sources/PackageGraph/PackageGraph+Loading.swift:615]: not sure if it makes a difference, but should this be inside the if condition in 616?

File: Tests/BuildTests/BuildPlanTests.swift [hunk 9]
```diff
@@ -2350,7 +2350,7 @@ final class BuildPlanTests: XCTestCase {
         result.checkProductsCount(1)
         result.checkTargetsCount(6)
 
-        XCTAssertTrue(result.targetMap.values.contains { $0.target.name == "Utils" && $0.target.moduleAliases?["Logging"] == "FooLogging" && $0.target.moduleAliases?["Math"] == "FooMath" })
+        XCTAssertTrue(result.targetMap.values.contains { $0.target.name == "Utils" && $0.target.moduleAliases == nil })
```
Comment (tomerd) [inline Tests/BuildTests/BuildPlanTests.swift:2353]: has this test changes because the behavior is now more correct than before?

File: Tests/BuildTests/BuildPlanTests.swift [hunk 10]
```diff
@@ -2268,7 +2268,7 @@ final class BuildPlanTests: XCTestCase {
         result.checkProductsCount(1)
         result.checkTargetsCount(4)
         
-        XCTAssertTrue(result.targetMap.values.contains { $0.target.name == "Utils" && $0.target.moduleAliases?["Logging"] == "FooLogging" })
+        XCTAssertTrue(result.targetMap.values.contains { $0.target.name == "Utils" && $0.target.moduleAliases == nil })
```
Comment (tomerd) [inline Tests/BuildTests/BuildPlanTests.swift:2271]: has this test changes because the behavior is now more correct than before?

Comment (elsh) [reply Tests/BuildTests/BuildPlanTests.swift:2271]: correct

File: Tests/BuildTests/BuildPlanTests.swift (hunk 9 above)
Comment (elsh) [reply Tests/BuildTests/BuildPlanTests.swift:2353]: yes

File: Sources/PackageGraph/PackageGraph+Loading.swift (hunk 8 above)
Comment (elsh) [reply Sources/PackageGraph/PackageGraph+Loading.swift:615]: no this is intentional; need to keep track of the entire chain as well as its target list -- sometimes module aliases are not defined in a middle upstream package, but still need to be propagated up if there are targets in upstream that match the alias defined downstream

**Review by tomerd — APPROVED** [review]

**Comment by elsh** [comment]
@swift-ci smoke test
